skbio/metadata/mixin.py

```diff
@@ -0,0 +1,519 @@
+           positional metadata behavior. These examples apply to all other
+           scikit-bio objects storing positional metadata.
+
+        >>> import pandas as pd
+        >>> from skbio import DNA
+        >>> seq = DNA('ACGT')
+        >>> seq.has_positional_metadata()
+        False
+        >>> seq = DNA('ACGT', positional_metadata=pd.DataFrame(index=range(4)))
+        >>> seq.has_positional_metadata()
+        False
+        >>> seq = DNA('ACGT', positional_metadata={'quality': range(4)})
+        >>> seq.has_positional_metadata()
+        True
+
+        """
+        return (self._positional_metadata is not None and
+                len(self.positional_metadata.columns) > 0)
+
+
+class IntervalMetadataMixin(with_metaclass(abc.ABCMeta, object)):
+    ''' Store metadata corresponding to Features and Intervals.
+
+    Parameters
+    ----------
+    features : dict of list
+        The dictionary of features and intervals where
+        the keys are hashble Feature objects and the values
+        are a list of intervals associated with each feature.
+
+    There are two underlying data structures namely
+    1. An IntervalTree (interval_metadata).
+    2. A hash table indexed by a Feature, where each entry contains a
+       list of Intervals(feature_metadata).
+
+    Attributes
+    ----------
+    interval_metadata
+    feature_metadata
+
+    >>> from skbio import DNA
+    >>> from skbio.metadata import Feature
+    >>> seq = DNA(
+    ...     'ACGT',
+    ...     positional_metadata={'quality': [3, 3, 20, 11],
+    ...                          'exons': [True, True, False, True]},
+    ...     interval_metadata={Feature(gene='hedgehog', location=0):[(0, 3)]})
+    >>> seq
+    DNA
+    -----------------------------
+    Positional metadata:
+        'exons': <dtype: bool>
+        'quality': <dtype: int64>
+    Interval metadata:
+        Number of features: 1
+        Number of intervals: 1
+    Stats:
+        length: 4
+        has gaps: False
+        has degenerates: False
+        has non-degenerates: True
+        GC-content: 50.00%
+    -----------------------------
+    0 ACGT
+    '''
+    def __init__(self, features=None):
+        pass
+
+    def _init_(self, features=None):
+        self.interval_metadata = IntervalMetadata(features=features)
+
+    @experimental(as_of="0.4.2")
+    def has_interval_metadata(self):
+        return len(self.interval_metadata.features) > 0
+
+        @abc.abstractmethod
+    def __eq__(self, other):
+        pass
+
+    def _eq_(self, other):
+        # We're not simply comparing self.positional_metadata to
+        # other.positional_metadata in order to avoid creating "empty"
+        # positional metadata representations on the objects if they don't have
+        # positional metadata.
+        if self.has_interval_metadata() and other.has_interval_metadata():
+            if not self.interval_metadata.equals(other.interval_metadata):
```

Comment:
ah, we can't postpone the implementation of `IntervalMetadata.__eq__` - it turns out needed for writing tests